### ISSUE TITLE
Enable apt repository for overlay packages

### ISSUE THREAD

**Issue by gboutry** [canonical/rockcraft#212]
## Use case

Install packages from a PPA while having its maintainerscripts executed

## Context
Using stage-packages directive I was able to do the following (minimal reproduction):
```yaml
name: keystone
summary: Openstack Keystone
license: Apache-2.0
description: |
  Ubuntu distribution of OpenStack Keystone
version: antelope

base: ubuntu:22.04
platforms:
  amd64:

parts:
  cloud-archive:
    plugin: nil
    build-packages:
      - software-properties-common
    override-pull: |
      add-apt-repository cloud-archive:antelope
      apt update
      craftctl default

  keystone:
    after: [cloud-archive]
    plugin: nil
    stage-packages:
      - keystone
```
Looking to the LXC build container, I have:
```bash
root@rockcraft-my-rock-name-52836265:~# find . -name "keystone*.deb"
./.cache/rockcraft/download/keystone-common_22.0.0+git2023011009.7d4047cb-0ubuntu1~cloud0_all.deb
./.cache/rockcraft/download/keystone_22.0.0+git2023011009.7d4047cb-0ubuntu1~cloud0_all.deb
./parts/keystone/stage_packages/keystone-common_22.0.0+git2023011009.7d4047cb-0ubuntu1~cloud0_all.deb
./parts/keystone/stage_packages/keystone_22.0.0+git2023011009.7d4047cb-0ubuntu1~cloud0_all.deb
```

These are the right version.

Trying the same with overlay step:

```diff 
diff repro-stage/rockcraft.yaml repro-overlay/rockcraft.yaml 
25c25
<     stage-packages:
---
>     overlay-packages:
```

Inside the LXC build container, I have:
```
root@rockcraft-my-rock-name-52836265:~# find . -name "keystone*.deb"
./overlay/packages/var/cache/apt/archives/partial/keystone-common_2%3a21.0.0-0ubuntu1_all.deb
./overlay/packages/var/cache/apt/archives/partial/keystone_2%3a21.0.0-0ubuntu1_all.deb
./overlay/packages/var/cache/apt/archives/keystone_2%3a21.0.0-0ubuntu1_all.deb
./overlay/packages/var/cache/apt/archives/keystone-common_2%3a21.0.0-0ubuntu1_all.deb
```

These are the distro versions you get from Jammy naturally.

From what I understand, what I do in `override-pull` updates the sources on the LXC container. When downloading packages in pull step for `overlay-packages`, the command is chrooted in the overlay and therefore does not have the modification in the sources.

Is there an easy way to modify the sources and have the cache updated in the overlay layer ?

## Modifying the overlay directly

<details>
  <summary>very wrong way</summary>

  ```yaml
  name: keystone
  summary: Openstack Keystone
  license: Apache-2.0
  description: |
    Ubuntu distribution of OpenStack Keystone
  version: antelope
  
  base: ubuntu:22.04
  platforms:
    amd64:
  
  parts:
    cloud-archive:
      plugin: nil
      overlay-packages:
        - software-properties-common
      overlay-script: |
        mount /etc/resolv.conf $CRAFT_OVERLAY/etc/resolv.conf --bind
        mount proc $CRAFT_OVERLAY/proc -t proc
        mount sysfs $CRAFT_OVERLAY/sys -t sysfs
        mount /dev $CRAFT_OVERLAY/dev --rbind --make-private
        chroot $CRAFT_OVERLAY stat /dev/null
        chroot $CRAFT_OVERLAY add-apt-repository cloud-archive:antelope
        chroot $CRAFT_OVERLAY apt update
        chroot $CRAFT_OVERLAY bash -c "DEBIAN_FRONTEND=noninteractive apt install apache2 keystone --yes"
        umount --recursive $CRAFT_OVERLAY/dev
        umount $CRAFT_OVERLAY/sys
        umount $CRAFT_OVERLAY/proc
        umount $CRAFT_OVERLAY/etc/resolv.conf
  ```
</details>

## Question

What's the expected way to enable a PPA inside an overlay layer ?
Is there a way from `override-pull` to modify apt sources lists meaningfully ?

**Comment by sergiusens** [comment]
@lengau is adding package-repositories support this pulse, but not for overlays, which is something we can see if can be done as part of this work or during our next 2 week stint.

**Comment by gboutry** [comment]
Hi @sergiusens, are there news about getting package repositories into overlays ?

**Comment by sergiusens** [comment]
@tigarmo did a spike which we will be working on in the coming pulses

**Comment by tigarmo** [comment]
This has now been fixed by #278
